src/compile/scale/range.ts

```diff
@@ -307,8 +312,64 @@ function defaultRange(channel: ScaleChannel, model: UnitModel): VgRange {
       // TODO: support custom rangeMin, rangeMax
       return [config.scale.minOpacity, config.scale.maxOpacity];
   }
-  /* istanbul ignore next: should never reach here */
-  throw new Error(`Scale range undefined for channel ${channel}`);
+}
+
+function getPositionStep(step: Step, model: UnitModel, channel: PositionScaleChannel): number| SignalRef {
+  const {encoding} = model;
+
+  const mergedScaleCmpt = model.getScaleComponent(channel);
+  const offsetChannel = getOffsetScaleChannel(channel);
+  const offsetDef = encoding[offsetChannel];
+  const stepFor = getStepFor({step, offsetIsDiscrete: isFieldDef(offsetDef) && isDiscrete(offsetDef.type)})
+
+  if (stepFor === 'offset' && channelHasField(encoding, offsetChannel)) {
+    const offsetScaleName = model.scaleName(offsetChannel);
+    const paddingInner = mergedScaleCmpt.get('paddingInner');
+    const padding = mergedScaleCmpt.get('padding');
+    return {signal: `${step.step} * bandspace(domain('${offsetScaleName}').length) / (1-${exprFromSignalRefOrValue(paddingInner ?? padding)})`}
+    ;
+  } else {
+    return step.step;
+  }
+}
+
+function getOffsetStep(step: Step, offsetScaleType: ScaleType) {
+  const stepFor = getStepFor({step, offsetIsDiscrete: hasDiscreteDomain(offsetScaleType)});
+  if (stepFor === 'offset') {
+    return {step: step.step};
+  }
+  return undefined;
+}
+
+function getOffsetRange(channel: string, model: UnitModel, offsetScaleType: ScaleType): VgRange {
+  const positionChannel = channel === XOFFSET ? 'x' : 'y';
+  const positionScaleCmpt = model.getScaleComponent(positionChannel);
+  const positionScaleType = positionScaleCmpt.get('type');
+  const positionScaleName=  model.scaleName(positionChannel);
+
+  if (positionScaleType === 'band') {
+    const size = getDiscretePositionSize(positionChannel, model.size, model.config.view);
+
+    if (isStep(size)) { // step is for offset
+      const step = getOffsetStep(size, offsetScaleType);
+      if (step) {
+        return step;
+      }
+    }
+    // otherwise use the position
+    return [0, {signal: `bandwidth('${positionScaleName}')`}];
+  } else { // continuous scale
+    throw new Error(`Cannot use ${channel} scale if ${positionChannel} scale is not discrete.`)
```

Comment:
We should never reach this line because we won't even generate the scale in the first place.